src/Witness/NTS.Witness/RPC/WitnessRpcClient.cs

```diff
@@ -0,0 +1,179 @@
+    }
+
+    public async Task<RpcInvokeResult<Dictionary<int, WitnessStartlist>>> RequestStartlistAsync()
+    {
+        return await _socket.InvokeOutputProcedure<Dictionary<int, WitnessStartlist>>(nameof(SendStartlist));
+    }
+
+    public async Task<RpcInvokeResult<IEnumerable<Participation>>> RequestParticipantsAsync()
+    {
+        return await _socket.InvokeOutputProcedure<IEnumerable<Participation>>(nameof(SendParticipants));
+    }
+
+    public Task ReceiveEntry(StartlistEntry entry, WitnessCollectionAction action)
+    {
+        var startlist = GetOrCreateStartlist(entry.PhaseNumber);
+        startlist.Update(entry, action);
+        return Task.CompletedTask;
+    }
+
+    public Task ReceiveEntryUpdate(Participation participation, WitnessCollectionAction action)
+    {
+        UpdateCollection(_participants, participation, action, p => p.Id);
+        return Task.CompletedTask;
+    }
+
+    public Task ReceiveWitnessEvent(WitnessSnapshotPayload payload)
+    {
+        if (payload?.Entries == null)
+        {
+            return Task.CompletedTask;
+        }
+
+        foreach (var snapshot in payload.Entries)
+        {
+            UpdateCollection(_snapshots, snapshot, WitnessCollectionAction.AddOrUpdate, s => s.Number);
+        }
+
+        return Task.CompletedTask;
+    }
+
+    public Task<Dictionary<int, WitnessStartlist>> SendStartlist()
+    {
+        EnsureStartlistsLoaded();
+        var copy = _startlists.ToDictionary(x => x.Key, x => new WitnessStartlist(x.Value));
+        return Task.FromResult(copy);
+    }
+
+    public Task<IEnumerable<Participation>> SendParticipants()
+    {
+        EnsureParticipantsLoaded();
+        return Task.FromResult<IEnumerable<Participation>>(_participants.ToList());
+    }
+
+    void EnsureStartlistsLoaded()
+    {
+        if (_startlists.Count > 0)
+        {
+            return;
+        }
+
+        var participations = DummyData.CreateParticipationsForStartlist();
+        var startList = new StartList(participations, () => { });
+        var entries = startList.Upcoming.Concat(startList.History);
+        foreach (var group in entries.GroupBy(entry => entry.PhaseNumber))
+        {
+            _startlists[group.Key] = new WitnessStartlist(group);
+        }
+    }
+
+    void EnsureParticipantsLoaded()
+    {
+        if (_participants.Count > 0)
+        {
+            return;
+        }
+
+        var participations = DummyData.CreateParticipationsForSnapshot(DEFAULT_SNAPSHOT_PARTICIPANTS);
+        _participants.AddRange(participations);
+    }
+
+    WitnessStartlist GetOrCreateStartlist(int phaseNumber)
+    {
+        if (!_startlists.TryGetValue(phaseNumber, out var startlist))
+        {
+            startlist = new WitnessStartlist();
+            _startlists[phaseNumber] = startlist;
+        }
+
+        return startlist;
+    }
+
+    static void UpdateCollection<T, TKey>(
```

Comment:
We should drop domain records, which will enable us to rely on consistent Equals behavior for domain objects. Once we have consistency we can create an extension method on `ICollection` for example which modifies the collection based on `action` type